src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -1254,6 +1254,46 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
                     _TryStopAutoScroll(ptr.PointerId());
                 }
             }
+            const auto terminalPos = _GetTerminalPosition(point.Position());
+            if (terminalPos != _lastHoveredCell)
+            {
+                const auto uri = _terminal->GetHyperlinkAtPosition(terminalPos);
+                if (!uri.empty())
+                {
+                    // Update the tooltip with the URI
+                    LinkTip().Content(winrt::box_value(uri));
+
+                    // Set the border thickness so it covers the entire cell
+                    const auto charSizeInPixels = CharacterDimensions();
+                    const auto htInDips = charSizeInPixels.Height / SwapChainPanel().CompositionScaleY();
+                    const auto wtInDips = charSizeInPixels.Width / SwapChainPanel().CompositionScaleX();
+                    const Thickness newThickness{ wtInDips, htInDips, 0, 0 };
+                    SubBorder().BorderThickness(newThickness);
+
+                    // Compute the location of the top left corner of the cell in DIPS
+                    const til::size marginsInDips{ til::math::rounding, GetPadding().Left, GetPadding().Top };
+                    const til::point startPos{ terminalPos.X, terminalPos.Y };
+                    const til::size fontSize{ _actualFont.GetSize() };
+                    const til::point posInPixels{ startPos * fontSize };
+                    const til::point posInDIPs{ posInPixels / SwapChainPanel().CompositionScaleX() };
+                    const til::point locationInDIPs{ posInDIPs + marginsInDips };
+
+                    // Move the border to the top left corner of the cell
+                    SubCanvas().SetLeft(SubBorder(), (locationInDIPs.x() - SwapChainPanel().ActualOffset().x));
+                    SubCanvas().SetTop(SubBorder(), (locationInDIPs.y() - SwapChainPanel().ActualOffset().y));
+                }
+                _lastHoveredCell = terminalPos;
+
+                const auto newId = _terminal->GetHyperlinkIdAtPosition(terminalPos);
+                // If the hyperlink ID changed, trigger a redraw all (so this will happen both when we move
+                // onto a link and when we move off a link)
+                if (newId != _lastHoveredId)
+                {
+                    _renderEngine->UpdateHyperlinkHoveredId(newId);
+                    _renderer->TriggerRedrawAll();
+                    _lastHoveredId = newId;
+                }

```

Comment:
I'm wondering if this logic (check equality, store last hovered ID, trigger redraw) all belongs inside `UpdateHyperlinkHoveredId`. Is that bad for the render engine contract @miniksa?

Reply:
(right now we're tracking the last ID in two places)

Reply:
Ugh its frustrating. The problem is that the specific engine has no way of signaling the renderer base to do the redraw all operation. When I originally laid it out, it was supposed to be getting commands inbound only.

This isn't that far off from us storing the invalid regions or cursor positions in multiple places. For instance, some of the renderers remember the last cursor pos they drew so they can erase the cursor "turds". Yes, it's two places... but there's precedent and I'm not sure what else is to be done beyond rearchitecting, which is out of scope.